add nodes and edges to completed event data

chriszirkel opened the pull request:
Hi!

Neovis does currently not support to receive the query results outside of the graph, for example to display the results as a table or list. In order to get the results, a second identical query would be necessary. 

I added nodes and edges to the event data object for `CompletionEvent `, so that event listeners can use this data directly without submitting a second query.

As this is my first contribution on GitHub, please let me know if I didn't follow the right proceudre and I will try to fix it! :-)

Best regards
Chris

thebestnom commented:
I think it's a good addition, but I think maybe adding a getter will be a better idea, that way you could also update the graph if you want to

Also I think that if you already touching the graph data and you don't mind, maybe adding the original nodes/edges to the data would be great

thebestnom commented:
Looks good, just indent correclty (you can use yarn/npm run eslint --fix)

chriszirkel commented:
Thank you for your quick reply and input! I added two getters for nodes and edges at the beginning of the class, right after the properties. I used the get syntax from JS. Let me know if that's what you meant :-)

Regarding your second point: with original nodes/edges you mean the raw results from neo4j? And where would you like me to add them?

thebestnom commented:
@jexp @johnymontana

thebestnom commented:
> Thank you for your quick reply and input! I added two getters for nodes and edges at the beginning of the class, right after the properties. I used the get syntax from JS. Let me know if that's what you meant :-)
> 
> Regarding your second point: with original nodes/edges you mean the raw results from neo4j? And where would you like me to add them?

Yeah, js getter are a good way to add that
And for the second point, yes I meant the original neo4j result (it's not exactly raw so I think it's a good thing to)
you can add that so when calling `buildNodeVisObject` or `buildEdgeVisObject` add the original as `original` or something on the object

Lastly, I just saw that you didn't update the dist files, how are you testing the code?

chriszirkel commented:
> > Thank you for your quick reply and input! I added two getters for nodes and edges at the beginning of the class, right after the properties. I used the get syntax from JS. Let me know if that's what you meant :-)
> > Regarding your second point: with original nodes/edges you mean the raw results from neo4j? And where would you like me to add them?
> 
> Yeah, js getter are a good way to add that
> And for the second point, yes I meant the original neo4j result (it's not exactly raw so I think it's a good thing to)
> you can add that so when calling `buildNodeVisObject` or `buildEdgeVisObject` add the original as `original` or something on the object
> 
> Lastly, I just saw that you didn't update the dist files, how are you testing the code?

Yes, you're right! I was honestly just figuring out how to build and test it. I did build for production and ran the tests. Let me know if I did it right? :-)

thebestnom commented:
Yeah! The test doesn't need tp build before hand, but doing them both are a good thing anyway

But I was mentioning about how you test your own changes, do you have a test project?
Also adding test is always a good way of knowing if your code works 😉

thebestnom commented:
Btw, I wouldn't have time to give more feedback today, but great work so far!

dukesun99 commented:
> > Thank you for your quick reply and input! I added two getters for nodes and edges at the beginning of the class, right after the properties. I used the get syntax from JS. Let me know if that's what you meant :-)
> > Regarding your second point: with original nodes/edges you mean the raw results from neo4j? And where would you like me to add them?
> 
> Yeah, js getter are a good way to add that
> And for the second point, yes I meant the original neo4j result (it's not exactly raw so I think it's a good thing to)
> you can add that so when calling `buildNodeVisObject` or `buildEdgeVisObject` add the original as `original` or something on the object
> 
> Lastly, I just saw that you didn't update the dist files, how are you testing the code?

@chriszirkel I also think adding the raw information of nodes and edges to vis network objects would be a good idea. Then after that, we can think of how to add events to allow users to define their own event handlers to handle events like click and double click.

chriszirkel commented:
> Yeah! The test doesn't need tp build before hand, but doing them both are a good thing anyway
> 
> But I was mentioning about how you test your own changes, do you have a test project?
> Also adding test is always a good way of knowing if your code works 😉

I have a project where I already use this libraray, but I coulnd't test it until I figured out that I need to build it and the copy the dist files over to my project. So before that, I was just expercting it to work :-D

I added two test steps to the existing cases which just compare the object length. Please let me know if that's sufficient.

I'm going to work on the other point tomorrow.

Thanks for your help and patience with me guys! :-)

chriszirkel commented:
> > > Thank you for your quick reply and input! I added two getters for nodes and edges at the beginning of the class, right after the properties. I used the get syntax from JS. Let me know if that's what you meant :-)
> > > Regarding your second point: with original nodes/edges you mean the raw results from neo4j? And where would you like me to add them?
> > 
> > 
> > Yeah, js getter are a good way to add that
> > And for the second point, yes I meant the original neo4j result (it's not exactly raw so I think it's a good thing to)
> > you can add that so when calling `buildNodeVisObject` or `buildEdgeVisObject` add the original as `original` or something on the object
> > Lastly, I just saw that you didn't update the dist files, how are you testing the code?
> 
> @chriszirkel I also think adding the raw information of nodes and edges to vis network objects would be a good idea. Then after that, we can think of how to add events to allow users to define their own event handlers to handle events like click and double click.

I added two fields for `neo4jNodes` and `neo4jEdges` - we could also name them `rawNodes` or something else.

The neo4j nodes and edges are added with node/edge identity as object key to the fields, just like the vis nodes/edges. The fields are being filled in `buildNodeVisObject`/`buildEdgeVisObject`. Maybe it's a good idea to extend the `_addNode` and `_addEdge`  methods with a second paramater for the neo4j object, so there's one single location where nodes and edges are being filled.

Let me know what you guys think!

thebestnom commented:
Any reason why not adding the raw data straight to the node?

chriszirkel commented:
Good point! I wasn’t thinking about that option, but it would definitely make more sense. Let me refactor it and I’ll come back to you.

thebestnom reviewed (approved):
Looks great!

chriszirkel commented:
I added the a property `raw` to node and edge and removed the neo4j fields. I added a new test case 'should save raw records' and tested it on my project. However, I noticed that the map `neovis._data.edges` is not being filled in testmode and therefore, I had to comment out this test step in the following lines:  https://github.com/neo4j-contrib/neovis.js/pull/127/commits/08eae26bf97068ab9edf61d2942781464c6c789c#diff-ade5369d026f310b1100d1b2340f6652R196-R197

Could you please take a look at this and see if it's due to the test moch setup?

thebestnom commented:
@jexp @johnymontana please merge

thebestnom commented:
> I added the a property `raw` to node and edge and removed the neo4j fields. I added a new test case 'should save raw records' and tested it on my project. However, I noticed that the map `neovis._data.edges` is not being filled in testmode and therefore, I had to comment out this test step in the following lines: [08eae26#diff-ade5369d026f310b1100d1b2340f6652R196-R197](https://github.com/neo4j-contrib/neovis.js/commit/08eae26bf97068ab9edf61d2942781464c6c789c#diff-ade5369d026f310b1100d1b2340f6652R196-R197)
> 
> Could you please take a look at this and see if it's due to the test moch setup?

This is wierd, it looks like it should work, a test above that looks like it working

thebestnom commented:
Oh, I know why 😂, `.get` get them by id, and the id counter is set between the relationship and nodes

chriszirkel commented:
> Oh, I know why 😂, `.get` get them by id, and the id counter is set between the relationship and nodes
> 
> Very easy to mistake that, just do `get(3)` in the edge
> 
> https://github.com/neo4j-contrib/neovis.js/blob/88fda0a34e0777692fea90dd76cc93b398209a7f/__tests__/testUtils.js#L6

Great, thanks! It works now!

thebestnom commented:
Almost forgot, do you mind adding the changes to the d.ts?

chriszirkel commented:
I added the properties `nodes` and `edges` to the type definition. I also defined two interfaces for `INode` and `IEdge`. Please have a look and let me know what you think.

thebestnom commented:
I'm pretty sure it uses vis data, so maybe just do that the getters return vis-data DataSet

chriszirkel commented:
You mean the type defintion or the getter itself?

thebestnom commented:
Type definition

chriszirkel commented:
> Type definition

If I understand correctly, you mean the defition of nodes and edges should return a vis-DataSet? But then, we would not have our self-defined `raw`property available on the object.

thebestnom commented:
Oh, my bad I forgot the difference between _nodes to _data.nodes, I think the getter should return the _data.nodes (and edges), and for the type definition, DataSet is generic, here an example use https://github.com/visjs/vis-network/blob/master/types/network/Network.d.ts

chriszirkel commented:
I changed the getters to return `_data.nodes` and `_data.edges`.

I would need some advice for the type definitions. I understand it should return a vis DataSet like here: https://github.com/visjs/vis-network/blob/master/types/network/Network.d.ts#L23

However, as we added a the `raw` property I guess the definiton must be enhanced somehow?

thebestnom commented:
I opened a PR to your branch instead of bugging you here

thebestnom commented:
I was added as maintainer 😄 I will fix your conflicts and then merge that when I'll have the time

thebestnom commented:
Tank you very much!
